### PR TITLE
enable graident checkpointing in DetaObjectDetection and add tests in Swin/Donut_Swin

### PR DESCRIPTION
**Pull request by SangbumChoi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by SangbumChoi** [comment]
I also found missing part that transfomers team did not put.

https://github.com/jozhang97/DETA/blob/f47e50efeb194357fb93a119f196a2e485fffdb5/models/deformable_transformer.py#L225

**Review by amyeroberts — COMMENTED** [review]
Amazing - thanks again for another great contribution! 

For the models, could you make sure that the tests for gradient checkpointing are enabled i.e. `def test_gradient_checkpointing(self):`?

**Comment by SangbumChoi** [comment]
```
root@1c4a6f3ab7da:/mnt/nas2/users/sbchoi/transformers# pytest tests/models/deta/
==================================================================== test session starts ====================================================================
platform linux -- Python 3.10.13, pytest-7.4.4, pluggy-1.0.0
rootdir: /mnt/nas2/users/sbchoi/transformers
configfile: pyproject.toml
plugins: hypothesis-6.92.0, anyio-4.2.0, hydra-core-1.3.2
collected 155 items

tests/models/deta/test_image_processing_deta.py ....ss........                                                                                        [  9%]
tests/models/deta/test_modeling_deta.py .........s...............ssssss.ssssssssss......s...............s......s.............sssssssssss.ssssssssssss [ 79%]
sss..s..s.......s.s.ss.sss....ss                                                                                                                      [100%]

===================================================================== warnings summary ======================================================================
../../../../../opt/conda/lib/python3.10/site-packages/_pytest/config/__init__.py:1373
  /opt/conda/lib/python3.10/site-packages/_pytest/config/__init__.py:1373: PytestConfigWarning: Unknown config option: doctest_glob

    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

src/transformers/deepspeed.py:23
  /mnt/nas2/users/sbchoi/transformers/src/transformers/deepspeed.py:23: FutureWarning: transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations
    warnings.warn(

../../../../../opt/conda/lib/python3.10/site-packages/torch/utils/cpp_extension.py:28
  /opt/conda/lib/python3.10/site-packages/torch/utils/cpp_extension.py:28: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    from pkg_resources import packaging  # type: ignore[attr-defined]

../../../../../opt/conda/lib/python3.10/site-packages/pkg_resources/__init__.py:2871
../../../../../opt/conda/lib/python3.10/site-packages/pkg_resources/__init__.py:2871
  /opt/conda/lib/python3.10/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('ruamel')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

tests/models/deta/test_modeling_deta.py::DetaModelTest::test_disk_offload_bin
  /opt/conda/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
    return self.fget.__get__(instance, owner)()

tests/models/deta/test_modeling_deta.py::DetaModelTest::test_fast_init_context_manager
  /mnt/nas2/users/sbchoi/transformers/tests/test_modeling_common.py:462: FutureWarning: `torch.testing.assert_allclose()` is deprecated since 1.12 and will be removed in a future release. Please use `torch.testing.assert_close()` instead. You can find detailed upgrade instructions in https://github.com/pytorch/pytorch/issues/61844.
    torch.testing.assert_allclose(init_instance.linear.bias, expected_bias, rtol=1e-3, atol=1e-4)

tests/models/deta/test_modeling_deta.py::DetaModelTest::test_fast_init_context_manager
  /mnt/nas2/users/sbchoi/transformers/tests/test_modeling_common.py:465: FutureWarning: `torch.testing.assert_allclose()` is deprecated since 1.12 and will be removed in a future release. Please use `torch.testing.assert_close()` instead. You can find detailed upgrade instructions in https://github.com/pytorch/pytorch/issues/61844.
    torch.testing.assert_allclose(

tests/models/deta/test_modeling_deta.py::DetaModelTest::test_gradient_checkpointing
tests/models/deta/test_modeling_deta.py::DetaModelTest::test_training_gradient_checkpointing
  /opt/conda/lib/python3.10/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.
    warnings.warn(

tests/models/deta/test_modeling_deta.py::DetaModelTest::test_model_outputs_equivalence
  /mnt/nas2/users/sbchoi/transformers/tests/test_modeling_common.py:1968: UserWarning: Use of index_put_ on expanded tensors is deprecated. Please clone() the tensor before performing this operation. This also applies to advanced indexing e.g. tensor[indices] = tensor (Triggered internally at /opt/conda/conda-bld/pytorch_1702400410390/work/aten/src/ATen/native/TensorAdvancedIndexing.cpp:708.)
    t[t != t] = 0

tests/models/deta/test_modeling_deta.py::DetaModelTest::test_model_outputs_equivalence
  /mnt/nas2/users/sbchoi/transformers/tests/test_modeling_common.py:1968: UserWarning: Use of masked_fill_ on expanded tensors is deprecated. Please clone() the tensor before performing this operation. This also applies to advanced indexing e.g. tensor[mask] = scalar (Triggered internally at /opt/conda/conda-bld/pytorch_1702400410390/work/aten/src/ATen/native/cuda/Indexing.cu:1564.)
    t[t != t] = 0

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= 96 passed, 59 skipped, 12 warnings in 56.64s ========================================================

```

**Comment by SangbumChoi** [comment]
```
root@1c4a6f3ab7da:/mnt/nas2/users/sbchoi/transformers# pytest tests/models/swin/
==================================================================== test session starts ====================================================================
platform linux -- Python 3.10.13, pytest-7.4.4, pluggy-1.0.0
rootdir: /mnt/nas2/users/sbchoi/transformers
configfile: pyproject.toml
plugins: hypothesis-6.92.0, anyio-4.2.0, hydra-core-1.3.2
collected 189 items

tests/models/swin/test_modeling_swin.py .........ssssss.ssssssssss...............s.......s.......ssssssss.sssssssssssssssssss.s........s.....sss....s [ 57%]
........                                                                                                                                              [ 61%]
tests/models/swin/test_modeling_tf_swin.py ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss                                   [100%]

===================================================================== warnings summary ======================================================================
../../../../../opt/conda/lib/python3.10/site-packages/_pytest/config/__init__.py:1373
  /opt/conda/lib/python3.10/site-packages/_pytest/config/__init__.py:1373: PytestConfigWarning: Unknown config option: doctest_glob

    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

src/transformers/deepspeed.py:23
  /mnt/nas2/users/sbchoi/transformers/src/transformers/deepspeed.py:23: FutureWarning: transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations
    warnings.warn(

../../../../../opt/conda/lib/python3.10/site-packages/torch/utils/cpp_extension.py:28
  /opt/conda/lib/python3.10/site-packages/torch/utils/cpp_extension.py:28: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    from pkg_resources import packaging  # type: ignore[attr-defined]

../../../../../opt/conda/lib/python3.10/site-packages/pkg_resources/__init__.py:2871
../../../../../opt/conda/lib/python3.10/site-packages/pkg_resources/__init__.py:2871
  /opt/conda/lib/python3.10/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('ruamel')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_fast_init_context_manager
  /mnt/nas2/users/sbchoi/transformers/tests/test_modeling_common.py:462: FutureWarning: `torch.testing.assert_allclose()` is deprecated since 1.12 and will be removed in a future release. Please use `torch.testing.assert_close()` instead. You can find detailed upgrade instructions in https://github.com/pytorch/pytorch/issues/61844.
    torch.testing.assert_allclose(init_instance.linear.bias, expected_bias, rtol=1e-3, atol=1e-4)

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_fast_init_context_manager
  /mnt/nas2/users/sbchoi/transformers/tests/test_modeling_common.py:465: FutureWarning: `torch.testing.assert_allclose()` is deprecated since 1.12 and will be removed in a future release. Please use `torch.testing.assert_close()` instead. You can find detailed upgrade instructions in https://github.com/pytorch/pytorch/issues/61844.
    torch.testing.assert_allclose(

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_fast_init_context_manager
  /opt/conda/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
    return self.fget.__get__(instance, owner)()

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_for_masked_image_modeling
tests/models/swin/test_modeling_swin.py::SwinModelTest::test_gradient_checkpointing
  /mnt/nas2/users/sbchoi/transformers/src/transformers/models/swin/modeling_swin.py:173: FutureWarning: logits attribute is deprecated and will be removed in version 5 of Transformers. Please use the reconstruction attribute to retrieve the final output instead.
    warnings.warn(

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_gradient_checkpointing
tests/models/swin/test_modeling_swin.py::SwinModelTest::test_training_gradient_checkpointing
  /opt/conda/lib/python3.10/site-packages/torch/utils/checkpoint.py:429: UserWarning: torch.utils.checkpoint: please pass in use_reentrant=True or use_reentrant=False explicitly. The default value of use_reentrant will be updated to be False in the future. To maintain current behavior, pass use_reentrant=True. It is recommended that you use use_reentrant=False. Refer to docs for more details on the differences between the two variants.
    warnings.warn(

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_pipeline_image_classification
  /opt/conda/lib/python3.10/site-packages/huggingface_hub/repocard.py:105: UserWarning: Repo card metadata block was not found. Setting CardData to empty.
    warnings.warn("Repo card metadata block was not found. Setting CardData to empty.")

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_torch_fx
  /opt/conda/lib/python3.10/site-packages/torch/overrides.py:110: UserWarning: 'has_cuda' is deprecated, please use 'torch.backends.cuda.is_built()'
    torch.has_cuda,

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_torch_fx
  /opt/conda/lib/python3.10/site-packages/torch/overrides.py:111: UserWarning: 'has_cudnn' is deprecated, please use 'torch.backends.cudnn.is_available()'
    torch.has_cudnn,

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_torch_fx
  /opt/conda/lib/python3.10/site-packages/torch/overrides.py:117: UserWarning: 'has_mps' is deprecated, please use 'torch.backends.mps.is_built()'
    torch.has_mps,

tests/models/swin/test_modeling_swin.py::SwinModelTest::test_torch_fx
  /opt/conda/lib/python3.10/site-packages/torch/overrides.py:118: UserWarning: 'has_mkldnn' is deprecated, please use 'torch.backends.mkldnn.is_available()'
    torch.has_mkldnn,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= 66 passed, 123 skipped, 17 warnings in 29.65s =======================================================
```

**Comment by SangbumChoi** [comment]
```
root@1c4a6f3ab7da:/mnt/nas2/users/sbchoi/transformers# pytest tests/models/donut/
==================================================================== test session starts ====================================================================
platform linux -- Python 3.10.13, pytest-7.4.4, pluggy-1.0.0
rootdir: /mnt/nas2/users/sbchoi/transformers
configfile: pyproject.toml
plugins: hypothesis-6.92.0, anyio-4.2.0, hydra-core-1.3.2
collected 117 items

tests/models/donut/test_image_processing_donut.py ...........                                                                                         [  9%]
tests/models/donut/test_modeling_donut_swin.py ........ssssss..sssssssss.....................s........sssssssssssssssssssssssssss.s........s.....sss. [ 96%]
...
```

**Comment by SangbumChoi** [comment]
@amyeroberts Added test function for each corresponding models!

File: src/transformers/models/deta/modeling_deta.py [hunk 1]
```diff
@@ -1771,6 +1779,7 @@ class DetaForObjectDetection(DetaPreTrainedModel):
     _tied_weights_keys = [r"bbox_embed\.\d+"]
     # We can't initialize the model on meta device as some weights are modified during the initialization
     _no_split_modules = None
+    supports_gradient_checkpointing = True
```
Comment (amyeroberts) [inline src/transformers/models/deta/modeling_deta.py:1782]: This should be set under the `DetaPretrainedModel` class. Once that's set there, then you can remove the custom implementations of `test_gradient_checkpointing`, as it will run the tests in `test_modeling_common` [here](https://github.com/huggingface/transformers/blob/f40b87de0ca234df61f76928956c4a2118c0b548/tests/test_modeling_common.py#L351)

**Review by amyeroberts — COMMENTED** [review]
Great - thanks for iterating on this and running the tests. Looking good! 

Just a small comment on the tests: setting `supports_gradient_checkpointing` in the right place should mean you can leverage the existing tests.

File: src/transformers/models/deta/modeling_deta.py (hunk 1 above)
Comment (SangbumChoi) [reply src/transformers/models/deta/modeling_deta.py]: @amyeroberts I didn't realise that! Thanks for letting me know this, it was so simple I also removed custom implementation also!

**Comment by SangbumChoi** [comment]
@amyeroberts I have fixed as you suggested. Overall it has been same as https://github.com/huggingface/transformers/pull/28686 with additional modification. Only modifying as the https://github.com/huggingface/transformers/pull/28686 will pass the CI but when we try in real code it will failure because there are gradient_checkpointing function in backbone.

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py
Comment (amyeroberts) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:1]: Let's remove the changes to DETR here as they'll be added in #28686. This way you both get your github contributions as you both worked on the respective models

**Review by amyeroberts — APPROVED** [review]
Great work - thanks for this contribution! 

Just a comment to remove the changes in DETR. Otherwise we're good to merge!

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py
Comment (SangbumChoi) [reply src/transformers/models/deformable_detr/modeling_deformable_detr.py:1]: @amyeroberts How about merging that PR first? In my personal view, it seems that there are no reason to remove `copied from flag` in DETA. I can resolve the conflict if it exist!

Comment (amyeroberts) [reply src/transformers/models/deformable_detr/modeling_deformable_detr.py:1]: Sounds good! Thanks :D

**Comment by SangbumChoi** [comment]
@amyeroberts I think it seems good after merging most recent PR?

File: src/transformers/models/deta/modeling_deta.py [hunk 2]
```diff
@@ -1040,6 +1041,7 @@ class DetaEncoder(DetaPreTrainedModel):
 
     def __init__(self, config: DetaConfig):
         super().__init__(config)
+        self.gradient_checkpointing = False
```
Comment (amyeroberts) [inline src/transformers/models/deta/modeling_deta.py:1044]: I think this shouldn't be here?

```suggestion
```

**Review by amyeroberts — APPROVED** [review]
Looks great. Just one comment and we're good to merge!

File: src/transformers/models/deta/modeling_deta.py (hunk 2 above)
Comment (SangbumChoi) [reply src/transformers/models/deta/modeling_deta.py]: I thought you were right. However, if you see the test CI if fails due to encoder also has override from the class PreTrainedModel.
@amyeroberts Can you confirm this? I might need to revert the last chanage!

Comment (amyeroberts) [reply src/transformers/models/deta/modeling_deta.py]: Yep - you're right, my bad!

**Comment by SangbumChoi** [comment]
@amyeroberts all fixed! but could you rerun the CI? It seems just CI got failed to load extension in torchvision

**Comment by amyeroberts** [comment]
@SangbumChoi Looks great! Yes, we had quite a few issues yesterday because of a new release of pytorch which broke everything 😢  A fix has been merged into main, which should resolve the version issues. Could you rebase? This should update the CI images and trigger another run.

**Comment by amyeroberts** [comment]
@SangbumChoi Thanks for iterating and another great contribution!
